components/suggestion/command_provider/command_provider.tsx

```diff
@@ -139,13 +141,29 @@ export default class CommandProvider extends Provider {
         const channel = this.isInRHS && selectedPost.channel_id ? getChannel(store.getState(), selectedPost.channel_id) : getCurrentChannel(store.getState());
 
         const args = {
-            channel_id: channel?.id,
+            channel_id: channel.id,
             ...(rootId && {root_id: rootId, parent_id: rootId}),
         };
 
-        Client4.getCommandAutocompleteSuggestionsList(command, teamId, args).then(
+        if (shouldHandleCommand(command, getCloudAppCommandLocations(store.getState()))) {
+            handleCommand(command, getCloudAppCommandLocations(store.getState()), args).then((matches => {
+                const terms = matches.map((suggestion) => suggestion.complete);
+                resultCallback({
+                    matchedPretext: command,
+                    terms,
+                    items: matches,
+                    component: CommandSuggestion,
+                });
+            }))
+            return;
+        }
+
+        const syncSuggestions = getSynchronousResults(pretext, getCloudAppCommandLocations(store.getState()));
```

Comment:
If I understand right, the shouldHandleCommand should cover all the cloudapp locations, right? If we return inside that if, why all other commands should trigger the synchronization?

Reply:
`shouldHandleCommand` returns true only if we are "inside" one of the App commands. Meaning, we have typed "/hello ", as opposed to a partial match such as:

- "/"
- "/hel"
"/hello"

Once we have a space typed after the base command, it is time for that command to take full control over the results. Otherwise, we need to make sure results for the base command show while we are still typing the base command. So the flow is:

if `shouldHandleCommand(text)`, that means we have typed at least "/hello ":
    - Give full control to the App command that matches the base command
else:
    - Fetch normal autocomplete items from the server
    - On client-side, merge in any results for matching against the App command base commands

Reply:
Should we consider augmenting the server-side of `getCommandAutocompleteSuggestionsList` to merge in the relevant app sub-commands, and thus simplifying the client code? Or is it too much work in comparison/impractical?

Reply:
Moving this to the server would also help the mobile case, no?

One trick we have not used yet but can (and should!?) consider is invoking plugin(s) from the server using the p.ServeHTTP hook, instead of defining new Hooks. Specifically, with plugin-apps it'd give us easy access to the plugin relative to adding new (public) Hook APIs.  I would consider the overhead for something like autocomplete to be minimal, ([de]/serializing and parsing HTTP request vs a dedicated Hook).